src/java.base/share/classes/java/net/Socket.java

```diff
@@ -518,57 +564,95 @@ private Socket(SocketAddress address, SocketAddress localAddr,
+    private static SocketImpl createImpl() {
+        SocketImplFactory factory = Socket.factory;
+        if (factory != null) {
+            return factory.createSocketImpl();
+        } else {
+            // create a SOCKS SocketImpl that delegates to a platform SocketImpl
+            SocketImpl delegate = SocketImpl.createPlatformSocketImpl(false);
+            return new SocksSocketImpl(delegate);
         }
     }
 
-    void setImpl(SocketImpl si) {
-         impl = si;
+    /**
+     * Returns the {@code SocketImpl} for this Socket, creating it, and the
+     * underlying socket, if required.
+     * @throws SocketException if creating the underlying socket fails
+     */
+    private SocketImpl getImpl() throws SocketException {
+        if ((state & SOCKET_CREATED) == 0) {
+            synchronized (socketLock) {
+                int s = state;   // re-read state
+                if ((s & SOCKET_CREATED) == 0) {
+                    if (isClosed(s)) {
+                        throw new SocketException("Socket is closed");
+                    }
+                    SocketImpl impl = this.impl;
+                    if (impl == null) {
+                        this.impl = impl = createImpl();
+                    }
+                    try {
+                        impl.create(true);
+                    } catch (SocketException e) {
+                        throw e;
+                    } catch (IOException e) {
+                        throw new SocketException(e.getMessage(), e);
+                    }
+                    getAndBitwiseOrState(SOCKET_CREATED);
+                }
+            }
+        }
+        return impl;
     }
 
     /**
-     * Sets impl to the system-default type of SocketImpl.
-     * @since 1.4
+     * Returns the SocketImpl, may be null.
      */
-    void setImpl() {
-        SocketImplFactory factory = Socket.factory;
-        if (factory != null) {
-            impl = factory.createSocketImpl();
-        } else {
-            // create a SOCKS SocketImpl that delegates to a platform SocketImpl
-            SocketImpl delegate = SocketImpl.createPlatformSocketImpl(false);
-            impl = new SocksSocketImpl(delegate);
+    SocketImpl impl() {
+        return impl;
+    }
+
+    /**
+     * Sets the SocketImpl. The SocketImpl is connected to a peer. The behavior for
+     * the case that the Socket was not a newly created Socket is unspecified. If
+     * there is an existing SocketImpl then it closed to avoid leaking resources.
+     * @throws SocketException if the Socket is closed
+     * @apiNote For ServerSocket use when accepting connections
+     */
+    void setConnectedImpl(SocketImpl si) throws SocketException {
+        synchronized (socketLock) {
+            if ((state & CLOSED) != 0) {
+                throw new SocketException("Socket is closed");
+            }
+            SocketImpl previous = impl;
+            impl = si;
+            state = (SOCKET_CREATED | BOUND | CONNECTED);
```

Comment:
Could we replace this line with `setConnected()` (which does the same thing as what's being done here)? The reason I say that is because it becomes a bit easier, in a IDE,  to see who all calls `setConnected()` and that helps understand which all places in the code trigger this state change. Given the context of this code, I suspect it wouldn't be any noticable impact on performance.
This is mostly a nit and you can ignore it if you prefer to keep it in the current form.

Reply:
In setConnectedImpl it needs to be very clear which state bits are set so I think I'd prefer to leave that as is.

Reply:
That's fine with me.